pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,661 @@
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        name,
+					ClusterName: parent.String(),
+				},
+				Spec: tenancyv1beta1.WorkspaceSpec{
+					Type: tenancyv1alpha1.ClusterWorkspaceTypeReference{
+						Name: tenancyv1alpha1.ClusterWorkspaceTypeName("Home"),
+						Path: tenancyv1alpha1.RootCluster.String(),
+					},
+				},
+				Status: tenancyv1beta1.WorkspaceStatus{
+					URL: homeWorkspaceURL.String(),
+				},
+			}
+		}
+
+		responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
+		return
+	}
+
+	needsAutomaticCreation, workspaceType := h.needsAutomaticCreation(workspaceName)
+	if needsAutomaticCreation {
+		isHome := workspaceType == "Home"
+		var retryAfterSeconds int
+		var retryLater bool
+		var localCheckError error
+		var foundLocally bool
+		foundLocally, retryLater, retryAfterSeconds, localCheckError = h.checkWorkspaceLocally(workspaceName, isHome, user.GetName())
+		if localCheckError != nil {
+			responsewriters.InternalError(rw, req, localCheckError)
+			return
+		}
+
+		if !foundLocally {
+			// Home or bucket workspace not found in the local informer
+			// Let's try to create it
+
+			// But first check we have the right to do so.
+			attributes := homeWorkspaceAuthorizerAttributes(user, "create")
+
+			if isHome && workspaceName != h.getHomeWorkspaceName(user.GetName()) {
+				// If we're checking a home workspace or home bucket workspace, but not of the consistent user, let's refuse.
+				responsewriters.Forbidden(ctx, attributes, rw, req, "home workspace can only be created by the user of the home workspace", homeWorkspaceCodecs)
+				return
+			}
+
+			// Test if the user has the right to create his Home workspace when it doesn't exists
+			// => test the create verb on the clusterworkspaces/workspace subresource named ~ in the root workspace.
+
+			if decision, reason, err := h.authz.Authorize(
+				request.WithCluster(ctx, request.Cluster{Name: tenancyv1alpha1.RootCluster}),
+				attributes); err != nil {
+				responsewriters.InternalError(rw, req, err)
+				return
+			} else if decision != authorizer.DecisionAllow {
+				responsewriters.Forbidden(ctx, attributes, rw, req, reason, homeWorkspaceCodecs)
+				return
+			}
+
+			var createError error
+			retryLater, retryAfterSeconds, createError = h.tryToCreate(ctx, user.GetName(), workspaceName, workspaceType)
+			if createError != nil {
+				responsewriters.ErrorNegotiated(createError, errorCodecs, schema.GroupVersion{}, rw, req)
+				return
+			}
+		}
+
+		if retryLater {
+			// Return a 429 status asking the client to try again after the creationDelay
+			rw.Header().Set("Retry-After", fmt.Sprintf("%d", retryAfterSeconds))
+			http.Error(rw, "Creating the home workspace", http.StatusTooManyRequests)
+			return
+		}
+	}
+	h.apiHandler.ServeHTTP(rw, req)
+}
+
+var reHomeWorkspaceChar = regexp.MustCompile("[^a-z0-9-]")
```

Comment:
go doc. What is this?

Reply:
Renamed and godoc'd in commit 3fa6113f512a3182c4f8712db50917fbdea6cafa